### PR TITLE
Issue 15 add GitHub actions

### LINKED ISSUE #15: Add GitHub Actions for CI

### PR DESCRIPTION
**Pull request by kan-fu**
Fix #15, #16. DMAS-77309.

You can check the CI result at https://github.com/OceanNetworksCanada/api-python-client/actions/runs/7147416348.

For tox, it is mainly set up for CI. Tox can help with the development, but is not required. I also added some documentations on how to contribute. I am not a native English speaker. Feel free to add comments if you find the documentation can be improved.

### CODE REVIEW

**Review by Jacob-Stevens-Haas — APPROVED** [review]
LGTM!  I understood all the documentation, and tox test/format/lint passes on my PC.  Your english is very clear.

**Comment by kan-fu** [comment]
I incorporated your commit from #13 to save you some time from rebasing. @Jacob-Stevens-Haas

File: tox.ini [hunk 1]
```diff
@@ -0,0 +1,51 @@
+[tox]
+min_version = 4.0
+env_list =
+    py{38,39,310,311,312}
+    format-check
+    lint
+isolated_build = True
```
Comment (Jacob-Stevens-Haas) [inline tox.ini:7]: Do you need to add `testenv or `testenv:format` to that list?

**Review by Jacob-Stevens-Haas — APPROVED** [review]
Hey Kan, thanks for making the CI PR!  I don't really know tox, so most of my comments are questions, the others are minor suggestions that you can feel free to ignore.

File: tox.ini [hunk 2]
```diff
@@ -0,0 +1,51 @@
+[tox]
```
[the remark this answers is not in the document]
Comment (kan-fu) [reply tox.ini:1]: As the [tox documentation](https://tox.wiki/en/latest/config.html#discovery-and-file-types) said, 

> Tox only supports ini-style. tox.ini and setup.cfg are by nature such files, while in pyproject.toml currently you can only inline the ini-style config.

The configuration looks a bit awkward in the pyproject.toml. That's why I think it is better to put it in tox.ini (some repos I checked that use tox.ini https://github.com/tox-dev/tox-gh, https://github.com/sphinx-contrib/napoleon, https://github.com/readthedocs/sphinx-autoapi).
```
[tool.tox]
legacy_tox_ini = """
    [tox]
    min_version = 4.0
    env_list =
        py310
        py39
        type

    [testenv]
    deps = pytest
    commands = pytest tests
"""
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      - name: Install tox
+        run: python -m pip install tox-gh>=1.2
+
+      - name: Setup test suite
+        run: tox --notest
+
+      - name: Run test suite
+        env:
+          TOKEN: ${{ secrets.TOKEN }}
+        run: tox --skip-pkg-install
```
[the remark this answers is not in the document]
Comment (kan-fu) [reply .github/workflows/ci.yml:49]: Some background information:

There are two popular tox github actions: [tox-gh](https://github.com/tox-dev/tox-gh) and [tox-gh-actions](https://github.com/ymyzk/tox-gh-actions). The second one was developed by an individual developer and was released earlier (2019 vs 2021). It is more popular considering the repo stars. The first one is developed by the tox development team. I chose the first one in this PR.

I copied the template from tox-gh, so I did not give too much thought about the implementation detail. In fact I see yml files from some [repos](https://github.com/readthedocs/sphinx-autoapi/blob/main/.github/workflows/main.yml#L28) that just run `tox`. I believe the tox dev team thinks separating the setup stage `tox --no-test` (line 43, 44) with the running stage `tox --skip-pkg-install` (line 46~49) is a better idea.  

For running just `tox`, it is different for github actions and local dev. For local run, it is defined in the `env_list` in the [tox.ini](https://tox.wiki/en/4.11.4/config.html#core), which in our case is py{38,39,310,311,312}, format-check, and lint. For github actions, I had some explanations in the [CONTRIBUTING.md](https://github.com/OceanNetworksCanada/api-python-client/blob/38cb0fc0dddbd3c61d59a74a5e9a4de1e38fb84a/CONTRIBUTING.md?plain=1#L128). But I think I need to modify it to make it more clear based on [tox-gh Readme](https://github.com/tox-dev/tox-gh#basic-example)

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+
+```shell
+# Sync the main first if you are on a fork.
+$ git fetch
+$ git rebase origin/main
+```
+
+### After making the pull request
+
+If you find that some GitHub checks failed, you can check the log in the Actions tab to see what went wrong.
```
[the remark this answers is not in the document]
Comment (kan-fu) [reply CONTRIBUTING.md:166]: Hmm, never knew that. Thanks for mentioning. I will add it.

I am not sure if an outside contributor has the permission to re-run failed jobs. I will keep that in mind and update the document when it happens

File: tox.ini (hunk 1 above)
Comment (kan-fu) [reply tox.ini:7]: [testenv](https://tox.wiki/en/4.11.4/user_guide.html#test-environments) section is like a super class, so it is not an environment. The user defined environments like `format` and `lint` override part (actually nearly all in our case) of the config in their own [testenv:XXX] section in tox.ini (they use the default [basepython](https://tox.wiki/en/4.11.4/config.html#python-options) version). pyNM does not have its own section, but it also inherits [testenv] and overrides the basepython option based on NM. 

The env_list is used when running `tox` without any arguments, it is not a declaration of what available environments we can use. Even though I delete line 4, I can still run `tox -e py38` given you have python 38 installed. In fact I am thinking I should only leave py38 locally and let github actions deal with the full version range test. 

I think most people use tox as a test (here it is pytest, format test, linting test). One big difference between `format` and `foramt-check` is that `format` will not raise an error when black and isort formats some files. That is why I have two environments for different purposes. Running `tox` (including the `format-check`) is usually the final step before making a pull request or a commit. 

I tend not to enforce the tox dev workflow. It is mainly used for setting up GitHub actions, but can be handy. Personally I treat tox as a command memo like package.json for npm. For dev workflow, we can run `black src test`, or `tox -e format`, or relying on integrated formatter plugin (that is me). Everyone can have their own workflow style, as long as the formatter, linter and pytest are happy when making a pull request.

**Comment by Jacob-Stevens-Haas** [comment]
Ah, thanks for all that info.  It (mostly) makes sense now :)

**Review by eliferguson — APPROVED** [review]

**Pull request merged**